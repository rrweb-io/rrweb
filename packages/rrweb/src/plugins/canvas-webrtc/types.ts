export interface WebRTCDataChannel {
  id: number;
}
