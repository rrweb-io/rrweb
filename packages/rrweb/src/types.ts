import {
  serializedNodeWithId,
  idNodeMap,
  INode,
  MaskInputOptions,
  SlimDOMOptions,
  MaskInputFn,
  MaskTextFn,
} from 'rrweb-snapshot';
import { PackFn, UnpackFn } from './packer/base';
import { IframeManager } from './record/iframe-manager';
import { ShadowDomManager } from './record/shadow-dom-manager';
import type { Replayer } from './replay';
import { CanvasManager } from './record/observers/canvas/canvas-manager';

export enum EventType {
  DomContentLoaded,
  Load,
  FullSnapshot,
  IncrementalSnapshot,
  Meta,
  Custom,
  Plugin,
}

export type domContentLoadedEvent = {
  type: EventType.DomContentLoaded;
  data: {};
};

export type loadedEvent = {
  type: EventType.Load;
  data: {};
};

export type fullSnapshotEvent = {
  type: EventType.FullSnapshot;
  data: {
    node: serializedNodeWithId;
    initialOffset: {
      top: number;
      left: number;
    };
  };
};

export type incrementalSnapshotEvent = {
  type: EventType.IncrementalSnapshot;
  data: incrementalData;
};

export type metaEvent = {
  type: EventType.Meta;
  data: {
    href: string;
    width: number;
    height: number;
  };
};

export type customEvent<T = unknown> = {
  type: EventType.Custom;
  data: {
    tag: string;
    payload: T;
  };
};

export type pluginEvent<T = unknown> = {
  type: EventType.Plugin;
  data: {
    plugin: string;
    payload: T;
  };
};

export type styleSheetEvent = {};

export enum IncrementalSource {
  Mutation,
  MouseMove,
  MouseInteraction,
  Scroll,
  ViewportResize,
  Input,
  TouchMove,
  MediaInteraction,
  StyleSheetRule,
  CanvasMutation,
  Font,
  Log,
  Drag,
  StyleDeclaration,
}

export type mutationData = {
  source: IncrementalSource.Mutation;
} & mutationCallbackParam;

export type mousemoveData = {
  source:
    | IncrementalSource.MouseMove
    | IncrementalSource.TouchMove
    | IncrementalSource.Drag;
  positions: mousePosition[];
};

export type mouseInteractionData = {
  source: IncrementalSource.MouseInteraction;
} & mouseInteractionParam;

export type scrollData = {
  source: IncrementalSource.Scroll;
} & scrollPosition;

export type viewportResizeData = {
  source: IncrementalSource.ViewportResize;
} & viewportResizeDimension;

export type inputData = {
  source: IncrementalSource.Input;
  id: number;
} & inputValue;

export type mediaInteractionData = {
  source: IncrementalSource.MediaInteraction;
} & mediaInteractionParam;

export type styleSheetRuleData = {
  source: IncrementalSource.StyleSheetRule;
} & styleSheetRuleParam;

export type styleDeclarationData = {
  source: IncrementalSource.StyleDeclaration;
} & styleDeclarationParam;

export type canvasMutationData = {
  source: IncrementalSource.CanvasMutation;
} & canvasMutationParam;

export type fontData = {
  source: IncrementalSource.Font;
} & fontParam;

export type incrementalData =
  | mutationData
  | mousemoveData
  | mouseInteractionData
  | scrollData
  | viewportResizeData
  | inputData
  | mediaInteractionData
  | styleSheetRuleData
  | canvasMutationData
  | fontData
  | styleDeclarationData;

export type event =
  | domContentLoadedEvent
  | loadedEvent
  | fullSnapshotEvent
  | incrementalSnapshotEvent
  | metaEvent
  | customEvent
  | pluginEvent;

export type eventWithTime = event & {
  timestamp: number;
  delay?: number;
};

export type blockClass = string | RegExp;

export type maskTextClass = string | RegExp;

export type SamplingStrategy = Partial<{
  /**
   * false means not to record mouse/touch move events
   * number is the throttle threshold of recording mouse/touch move
   */
  mousemove: boolean | number;
  /**
   * number is the throttle threshold of mouse/touch move callback
   */
  mousemoveCallback: number;
  /**
   * false means not to record mouse interaction events
   * can also specify record some kinds of mouse interactions
   */
  mouseInteraction: boolean | Record<string, boolean | undefined>;
  /**
   * number is the throttle threshold of recording scroll
   */
  scroll: number;
  /**
   * number is the throttle threshold of recording media interactions
   */
  media: number;
  /**
   * 'all' will record all the input events
   * 'last' will only record the last input value while input a sequence of chars
   */
  input: 'all' | 'last';
}>;

export type RecordPlugin<TOptions = unknown> = {
  name: string;
  observer?: (cb: Function, win: IWindow, options: TOptions) => listenerHandler;
  eventProcessor?: <TExtend>(event: eventWithTime) => eventWithTime & TExtend;
  options: TOptions;
};

export type recordOptions<T> = {
  emit?: (e: T, isCheckout?: boolean) => void;
  checkoutEveryNth?: number;
  checkoutEveryNms?: number;
  blockClass?: blockClass;
  blockSelector?: string;
  ignoreClass?: string;
  maskTextClass?: maskTextClass;
  maskTextSelector?: string;
  maskAllInputs?: boolean;
  maskInputOptions?: MaskInputOptions;
  maskInputFn?: MaskInputFn;
  maskTextFn?: MaskTextFn;
  slimDOMOptions?: SlimDOMOptions | 'all' | true;
  inlineStylesheet?: boolean;
  hooks?: hooksParam;
  packFn?: PackFn;
  sampling?: SamplingStrategy;
  recordCanvas?: boolean;
  userTriggeredOnInput?: boolean;
  collectFonts?: boolean;
  inlineImages?: boolean;
  plugins?: RecordPlugin[];
  // departed, please use sampling options
  mousemoveWait?: number;
  keepIframeSrcFn?: KeepIframeSrcFn;
};

export type observerParam = {
  mutationCb: mutationCallBack;
  mousemoveCb: mousemoveCallBack;
  mouseInteractionCb: mouseInteractionCallBack;
  scrollCb: scrollCallback;
  viewportResizeCb: viewportResizeCallback;
  inputCb: inputCallback;
  mediaInteractionCb: mediaInteractionCallback;
  blockClass: blockClass;
  blockSelector: string | null;
  ignoreClass: string;
  maskTextClass: maskTextClass;
  maskTextSelector: string | null;
  maskInputOptions: MaskInputOptions;
  maskInputFn?: MaskInputFn;
  maskTextFn?: MaskTextFn;
  inlineStylesheet: boolean;
  styleSheetRuleCb: styleSheetRuleCallback;
  styleDeclarationCb: styleDeclarationCallback;
  canvasMutationCb: canvasMutationCallback;
  fontCb: fontCallback;
  sampling: SamplingStrategy;
  recordCanvas: boolean;
  inlineImages: boolean;
  userTriggeredOnInput: boolean;
  collectFonts: boolean;
  slimDOMOptions: SlimDOMOptions;
  doc: Document;
  mirror: Mirror;
  iframeManager: IframeManager;
  shadowDomManager: ShadowDomManager;
  canvasManager: CanvasManager;
  plugins: Array<{
    observer: Function;
    callback: Function;
    options: unknown;
  }>;
};

export type MutationBufferParam = Pick<
  observerParam,
  | 'mutationCb'
  | 'blockClass'
  | 'blockSelector'
  | 'maskTextClass'
  | 'maskTextSelector'
  | 'inlineStylesheet'
  | 'maskInputOptions'
  | 'maskTextFn'
  | 'maskInputFn'
  | 'recordCanvas'
  | 'inlineImages'
  | 'slimDOMOptions'
  | 'doc'
  | 'mirror'
  | 'iframeManager'
  | 'shadowDomManager'
  | 'canvasManager'
>;

export type hooksParam = {
  mutation?: mutationCallBack;
  mousemove?: mousemoveCallBack;
  mouseInteraction?: mouseInteractionCallBack;
  scroll?: scrollCallback;
  viewportResize?: viewportResizeCallback;
  input?: inputCallback;
  mediaInteaction?: mediaInteractionCallback;
  styleSheetRule?: styleSheetRuleCallback;
  styleDeclaration?: styleDeclarationCallback;
  canvasMutation?: canvasMutationCallback;
  font?: fontCallback;
};

// https://dom.spec.whatwg.org/#interface-mutationrecord
export type mutationRecord = {
  type: string;
  target: Node;
  oldValue: string | null;
  addedNodes: NodeList;
  removedNodes: NodeList;
  attributeName: string | null;
};

export type textCursor = {
  node: Node;
  value: string | null;
};
export type textMutation = {
  id: number;
  value: string | null;
};

export type styleAttributeValue = {
  [key: string]: styleValueWithPriority | string | false;
};

export type styleValueWithPriority = [string, string];

export type attributeCursor = {
  node: Node;
  attributes: {
    [key: string]: string | styleAttributeValue | null;
  };
};
export type attributeMutation = {
  id: number;
  attributes: {
    [key: string]: string | styleAttributeValue | null;
  };
};

export type removedNodeMutation = {
  parentId: number;
  id: number;
  isShadow?: boolean;
};

export type addedNodeMutation = {
  parentId: number;
  // Newly recorded mutations will not have previousId any more, just for compatibility
  previousId?: number | null;
  nextId: number | null;
  node: serializedNodeWithId;
};

export type mutationCallbackParam = {
  texts: textMutation[];
  attributes: attributeMutation[];
  removes: removedNodeMutation[];
  adds: addedNodeMutation[];
  isAttachIframe?: true;
};

export type mutationCallBack = (m: mutationCallbackParam) => void;

export type mousemoveCallBack = (
  p: mousePosition[],
  source:
    | IncrementalSource.MouseMove
    | IncrementalSource.TouchMove
    | IncrementalSource.Drag,
) => void;

export type mousePosition = {
  x: number;
  y: number;
  id: number;
  timeOffset: number;
};

export type mouseMovePos = {
  x: number;
  y: number;
  id: number;
  debugData: incrementalData;
};

export enum MouseInteractions {
  MouseUp,
  MouseDown,
  Click,
  ContextMenu,
  DblClick,
  Focus,
  Blur,
  TouchStart,
  TouchMove_Departed, // we will start a separate observer for touch move event
  TouchEnd,
  TouchCancel,
}

export enum CanvasContext {
  '2D',
  WebGL,
  WebGL2,
}

export type SerializedWebGlArg =
  | {
      rr_type: 'ArrayBuffer';
      base64: string; // base64
    }
  | {
      rr_type: string;
      src: string; // url of image
    }
  | {
      rr_type: string;
      args: SerializedWebGlArg[];
    }
  | {
      rr_type: string;
      index: number;
    }
  | string
  | number
  | boolean
  | null
  | SerializedWebGlArg[];

type mouseInteractionParam = {
  type: MouseInteractions;
  id: number;
  x: number;
  y: number;
};

export type mouseInteractionCallBack = (d: mouseInteractionParam) => void;

export type scrollPosition = {
  id: number;
  x: number;
  y: number;
};

export type scrollCallback = (p: scrollPosition) => void;

export type styleSheetAddRule = {
  rule: string;
  index?: number | number[];
};

export type styleSheetDeleteRule = {
  index: number | number[];
};

export type styleSheetRuleParam = {
  id: number;
  removes?: styleSheetDeleteRule[];
  adds?: styleSheetAddRule[];
};

export type styleSheetRuleCallback = (s: styleSheetRuleParam) => void;

export type styleDeclarationParam = {
  id: number;
  index: number[];
  set?: {
    property: string;
    value: string | null;
    priority: string | undefined;
  };
  remove?: {
    property: string;
  };
};

export type styleDeclarationCallback = (s: styleDeclarationParam) => void;

export type canvasMutationCommand = {
  property: string;
  args: Array<unknown>;
  setter?: true;
};

export type canvasMutationParam =
  | {
      id: number;
      type: CanvasContext;
      commands: canvasMutationCommand[];
    }
  | ({
      id: number;
      type: CanvasContext;
    } & canvasMutationCommand);

export type canvasMutationWithType = {
  type: CanvasContext;
} & canvasMutationCommand;

export type canvasMutationCallback = (p: canvasMutationParam) => void;

export type canvasManagerMutationCallback = (
  target: HTMLCanvasElement,
  p: canvasMutationWithType,
) => void;

export type fontParam = {
  family: string;
  fontSource: string;
  buffer: boolean;
  descriptors?: FontFaceDescriptors;
};

export type fontCallback = (p: fontParam) => void;

export type viewportResizeDimension = {
  width: number;
  height: number;
};

export type viewportResizeCallback = (d: viewportResizeDimension) => void;

export type inputValue = {
  text: string;
  isChecked: boolean;

  // `userTriggered` indicates if this event was triggered directly by user (userTriggered: true)
  // or was triggered indirectly (userTriggered: false)
  // Example of `userTriggered` in action:
  // User clicks on radio element (userTriggered: true) which triggers the other radio element to change (userTriggered: false)
  userTriggered?: boolean;
};

export type inputCallback = (v: inputValue & { id: number }) => void;

export const enum MediaInteractions {
  Play,
  Pause,
  Seeked,
  VolumeChange,
}

export type mediaInteractionParam = {
  type: MediaInteractions;
  id: number;
  currentTime?: number;
  volume?: number;
  muted?: boolean;
};

export type mediaInteractionCallback = (p: mediaInteractionParam) => void;

export type DocumentDimension = {
  x: number;
  y: number;
  // scale value relative to its parent iframe
  relativeScale: number;
  // scale value relative to the root iframe
  absoluteScale: number;
};

export type Mirror = {
  map: idNodeMap;
  getId: (n: INode) => number;
  getNode: (id: number) => INode | null;
  removeNodeFromMap: (n: INode) => void;
  has: (id: number) => boolean;
  reset: () => void;
};

export type throttleOptions = {
  leading?: boolean;
  trailing?: boolean;
};

export type listenerHandler = () => void;
export type hookResetter = () => void;

export type ReplayPlugin = {
  handler: (
    event: eventWithTime,
    isSync: boolean,
    context: { replayer: Replayer },
  ) => void;
};
export type playerConfig = {
  speed: number;
  maxSpeed: number;
  root: Element;
  loadTimeout: number;
  skipInactive: boolean;
  showWarning: boolean;
  showDebug: boolean;
  blockClass: string;
  liveMode: boolean;
  insertStyleRules: string[];
  triggerFocus: boolean;
  UNSAFE_replayCanvas: boolean;
  pauseAnimation?: boolean;
  mouseTail:
    | boolean
    | {
        duration?: number;
        lineCap?: string;
        lineWidth?: number;
        strokeStyle?: string;
      };
  unpackFn?: UnpackFn;
  plugins?: ReplayPlugin[];
};

export type playerMetaData = {
  startTime: number;
  endTime: number;
  totalTime: number;
};

export type missingNode = {
  node: Node;
  mutation: addedNodeMutation;
};
export type missingNodeMap = {
  [id: number]: missingNode;
};

export type actionWithDelay = {
  doAction: () => void;
  delay: number;
};

export type Handler = (event?: unknown) => void;

export type Emitter = {
  on(type: string, handler: Handler): void;
  emit(type: string, event?: unknown): void;
  off(type: string, handler: Handler): void;
};

export type Arguments<T> = T extends (...payload: infer U) => unknown
  ? U
  : unknown;

export enum ReplayerEvents {
  Start = 'start',
  Pause = 'pause',
  Resume = 'resume',
  Resize = 'resize',
  Finish = 'finish',
  FullsnapshotRebuilded = 'fullsnapshot-rebuilded',
  LoadStylesheetStart = 'load-stylesheet-start',
  LoadStylesheetEnd = 'load-stylesheet-end',
  SkipStart = 'skip-start',
  SkipEnd = 'skip-end',
  MouseInteraction = 'mouse-interaction',
  EventCast = 'event-cast',
  CustomEvent = 'custom-event',
  Flush = 'flush',
  StateChange = 'state-change',
  PlayBack = 'play-back',
}

// store the state that would be changed during the process(unmount from dom and mount again)
export type ElementState = {
  // [scrollLeft,scrollTop]
  scroll?: [number, number];
};

export type KeepIframeSrcFn = (src: string) => boolean;

declare global {
  interface Window {
    FontFace: typeof FontFace;
  }
}

export type IWindow = Window & typeof globalThis;

export type Optional<T, K extends keyof T> = Pick<Partial<T>, K> & Omit<T, K>;
