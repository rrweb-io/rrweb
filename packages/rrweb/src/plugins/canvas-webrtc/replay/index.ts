import type { RRNode } from 'rrdom';
import type { Mirror } from 'rrweb-snapshot';
import SimplePeer from 'simple-peer-light';
import type { Replayer } from '../../../replay';
import type { ReplayPlugin } from '../../../types';
import type { WebRTCDataChannel } from '../types';

// TODO: restrict callback to real nodes only, or make sure callback gets called when real node gets added to dom as well

export class RRWebPluginCanvasWebRTCReplay {
  private canvasFoundCallback: (
    node: Node | RRNode,
    context: { id: number; replayer: Replayer },
  ) => void;
  private signalSendCallback: (signal: RTCSessionDescriptionInit) => void;
  private mirror: Mirror;

  constructor({
    canvasFoundCallback,
    signalSendCallback,
  }: {
    canvasFoundCallback: RRWebPluginCanvasWebRTCReplay['canvasFoundCallback'];
    signalSendCallback: RRWebPluginCanvasWebRTCReplay['signalSendCallback'];
  }) {
    this.canvasFoundCallback = canvasFoundCallback;
    this.signalSendCallback = signalSendCallback;
  }

  public initPlugin(): ReplayPlugin {
    return {
      onBuild: (
        node: Node | RRNode,
        context: { id: number; replayer: Replayer },
      ) => {
        if (node.nodeName === 'CANVAS') {
          this.canvasFoundCallback(node, context);
        }
      },
      getMirror: (mirror: Mirror) => {
        this.mirror = mirror;
      },
    };
  }

  private startStream(
    target: HTMLCanvasElement | HTMLVideoElement,
    stream: MediaStream,
  ) {
    if (this.runningStreams.has(stream)) return;

    if (target.tagName === 'VIDEO') {
      const remoteVideo = target as HTMLVideoElement;
      remoteVideo.srcObject = stream;
      void remoteVideo.play();
      this.runningStreams.add(stream);
      return;
    }

    if ('MediaStreamTrackProcessor' in window) {
      const canvas = target as HTMLCanvasElement;
      const ctx = canvas.getContext('2d');
      if (!ctx)
        throw new Error(
          `startStream: Could not get 2d canvas context for ${canvas.outerHTML}`,
        );
      const track = stream.getVideoTracks()[0]; // MediaStream.getVideoTracks()[0]
      const processor = new MediaStreamTrackProcessor({ track: track });
      const reader = processor.readable.getReader();
      const readChunk = function () {
        void reader.read().then(({ done, value }) => {
          if (!value) return;
          // the MediaStream video can have dynamic size based on bandwidth available
          if (
            canvas.width !== value.displayWidth ||
            canvas.height !== value.displayHeight
          ) {
            canvas.width = value.displayWidth;
            canvas.height = value.displayHeight;
          }
          ctx.clearRect(0, 0, canvas.width, canvas.height);
          // value is a VideoFrame
          ctx.drawImage(value, 0, 0);
          value.close(); // close the VideoFrame when we're done with it
          if (!done) {
            readChunk();
          }
        });
      };
      readChunk();
      this.runningStreams.add(stream);
    } else {
      // Fallback for non-Chromium browsers.
      // Replaces the canvas element with a video element.

      const remoteVideo = document.createElement('video');
      remoteVideo.setAttribute('autoplay', 'true');
      remoteVideo.setAttribute('playsinline', 'true');

      // const { id } = mutation;
      remoteVideo.setAttribute('width', target.width.toString());
      remoteVideo.setAttribute('height', target.height.toString());
      target.replaceWith(remoteVideo);

      this.startStream(remoteVideo, stream);
    }
  }

  private peer: SimplePeer.Instance | null = null;
  private streamNodeMap = new Map<string, number>();
  private streams = new Set<MediaStream>();
  private runningStreams = new WeakSet<MediaStream>();
  public signalReceive(msg: RTCSessionDescriptionInit) {
    if (!this.peer) {
      this.peer = new SimplePeer({
        initiator: false,
        // trickle: false,
      });

      this.peer.on('error', (err: Error) => {
        this.peer = null;
        console.log('error', err);
      });

      this.peer.on('close', () => {
        this.peer = null;
        console.log('closing');
      });

      this.peer.on('signal', (data: RTCSessionDescriptionInit) => {
        this.signalSendCallback(data);
      });

      this.peer.on('connect', () => {
        // connected!
      });

      this.peer.on('data', (data: SimplePeer.SimplePeerData) => {
        try {
          const json = JSON.parse(data as string) as WebRTCDataChannel;
          this.streamNodeMap.set(json.streamId, json.nodeId);
        } catch (error) {
          console.error('Could not parse data', error);
        }
        this.flushStreams();
      });

      this.peer.on('stream', (stream: MediaStream) => {
        this.streams.add(stream);
        this.flushStreams();
      });
    }
    this.peer.signal(msg);
  }

  private flushStreams() {
    this.streams.forEach((stream) => {
      const nodeId = this.streamNodeMap.get(stream.id);
      if (!nodeId) return;
      const target = this.mirror.getNode(nodeId) as
        | HTMLCanvasElement
        | HTMLVideoElement
        | null;
      // got remote video stream, now let's show it in a video or canvas element
      if (target) this.startStream(target, stream);
    });
  }
}
