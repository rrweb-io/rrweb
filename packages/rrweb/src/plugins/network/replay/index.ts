import { NetworkData, NETWORK_PLUGIN_NAME } from '../record';
import type { eventWithTime } from '@rrweb/types';
import { EventType } from '@rrweb/types';
import type { ReplayPlugin } from '../../../types';

export type OnNetworkData = (data: NetworkData) => void;

export type NetworkReplayOptions = {
  onNetworkData: OnNetworkData;
};

export const getReplayNetworkPlugin: (
  options: NetworkReplayOptions,
) => ReplayPlugin = (options) => {
  return {
    handler(event: eventWithTime) {
      if (
        event.type === EventType.Plugin &&
        event.data.plugin === NETWORK_PLUGIN_NAME
      ) {
        const networkData = event.data.payload as NetworkData;
        options.onNetworkData(networkData);
      }
    },
  };
};
