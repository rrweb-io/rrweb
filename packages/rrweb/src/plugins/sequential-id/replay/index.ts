import type { SequentialIdOptions } from '../record';
import { ReplayPlugin, eventWithTime } from '../../../types';

type Options = SequentialIdOptions & {
  warnOnMissingId: boolean;
};

const defaultOptions: Options = {
  key: '_sid',
  warnOnMissingId: true,
};

export const getReplaySequentialIdPlugin: (
  options?: Partial<Options>,
) => ReplayPlugin = (options) => {
  const { key, warnOnMissingId } = options
    ? Object.assign({}, defaultOptions, options)
    : defaultOptions;
  let currentId = 1;

  return {
    handler(event: eventWithTime) {
      if (key in event) {
        const id = ((event as unknown) as Record<string, number>)[key];
        if (id !== currentId) {
          console.error(
            `[sequential-id-plugin]: expect to get an id with value "${currentId}", but got "${id}"`,
          );
        } else {
          currentId++;
        }
      } else if (warnOnMissingId) {
        console.warn(
          `[sequential-id-plugin]: failed to get id in key: "${key}"`,
        );
      }
    },
  };
};
