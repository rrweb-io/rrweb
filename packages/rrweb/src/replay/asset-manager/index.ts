import type {
  RebuildAssetManagerFinalStatus,
  RebuildAssetManagerInterface,
  RebuildAssetManagerStatus,
  assetEvent,
  captureAssetsParam,
} from '@rrweb/types';
import { deserializeArg } from '../canvas/deserialize-args';
import { getSourcesFromSrcset, isAttributeCacheable } from 'rrweb-snapshot';
import type { RRElement } from 'rrdom';
import { updateSrcset } from './update-srcset';

export default class AssetManager implements RebuildAssetManagerInterface {
  private originalToObjectURLMap: Map<string, string> = new Map();
  private nodeIdAttributeHijackedMap: Map<number, Map<string, string>> =
    new Map();
  private loadingURLs: Set<string> = new Set();
  private failedURLs: Set<string> = new Set();
  private callbackMap: Map<
    string,
    Array<(status: RebuildAssetManagerFinalStatus) => void>
  > = new Map();
  private config: captureAssetsParam | undefined;
  private liveMode: boolean;

  constructor(
    { liveMode }: { liveMode: boolean },
    config?: captureAssetsParam | undefined,
  ) {
    this.liveMode = liveMode;
    this.config = config;
  }

  public async add(event: assetEvent) {
    const { data } = event;
    const { url, payload, failed } = { payload: false, failed: false, ...data };
    if (failed) {
      this.failedURLs.add(url);
      this.executeCallbacks(url, { status: 'failed' });
      return;
    }
    this.loadingURLs.add(url);

    // tracks if deserializing did anything, not really needed for AssetManager
    const status = {
      isUnchanged: true,
    };

    // TODO: extract the logic only needed for assets from deserializeArg
    const result = (await deserializeArg(new Map(), null, status)(payload)) as
      | Blob
      | MediaSource;

    const objectURL = URL.createObjectURL(result);
    this.originalToObjectURLMap.set(url, objectURL);
    this.loadingURLs.delete(url);
    this.executeCallbacks(url, { status: 'loaded', url: objectURL });
  }

  private executeCallbacks(
    url: string,
    status: RebuildAssetManagerFinalStatus,
  ) {
    const callbacks = this.callbackMap.get(url);
    while (callbacks && callbacks.length > 0) {
      const callback = callbacks.pop();
      if (!callback) {
        break;
      }
      callback(status);
    }
  }

  // TODO: turn this into a true promise that throws if the asset fails to load
  public async whenReady(url: string): Promise<RebuildAssetManagerFinalStatus> {
    const currentStatus = this.get(url);
    if (
      currentStatus.status === 'loaded' ||
      currentStatus.status === 'failed'
    ) {
      return currentStatus;
    }
    let resolve: (status: RebuildAssetManagerFinalStatus) => void;
    const promise = new Promise<RebuildAssetManagerFinalStatus>((r) => {
      resolve = r;
    });
    if (!this.callbackMap.has(url)) {
      this.callbackMap.set(url, []);
    }
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    this.callbackMap.get(url)!.push(resolve!);

    return promise;
  }

  public get(url: string): RebuildAssetManagerStatus {
    const result = this.originalToObjectURLMap.get(url);

    if (result) {
      return {
        status: 'loaded',
        url: result,
      };
    }

    if (this.loadingURLs.has(url)) {
      return {
        status: 'loading',
      };
    }

    if (this.failedURLs.has(url)) {
      return {
        status: 'failed',
      };
    }

    return {
      status: 'unknown',
    };
  }

  public isCacheable(
    n: RRElement | Element,
    attribute: string,
    value: string,
  ): boolean {
    if (!isAttributeCacheable(n as Element, attribute)) return false;

    if (attribute === 'srcset') {
      return getSourcesFromSrcset(value).some((source) =>
        this.isURLOfCacheableOrigin(source),
      );
    } else {
      return this.isURLOfCacheableOrigin(value);
    }
  }

  public isURLOfCacheableOrigin(url: string): boolean {
    if (url.startsWith('data:')) return false;

    const { origins: cachedOrigins = false, objectURLs = false } =
      this.config || {};
    if (objectURLs && url.startsWith(`blob:`)) {
      return true;
    }

    if (Array.isArray(cachedOrigins)) {
      try {
        const { origin } = new URL(url);
        return cachedOrigins.some((o) => o === origin);
      } catch {
        return false;
      }
    }

    return cachedOrigins;
  }

  public async manageAttribute(
    node: RRElement | Element,
    nodeId: number,
    attribute: string,
  ): Promise<unknown> {
    const originalValue = node.getAttribute(attribute);
    if (node.nodeName === 'IMG')
      console.log(
        'AssetManager.manageAttribute',
        node.nodeName,
        attribute,
        originalValue,
        'livemode',
        this.liveMode,
      );
    if (!originalValue || !this.isCacheable(node, attribute, originalValue))
      return false;

    const promises: Promise<unknown>[] = [];

    if (attribute === 'srcset') {
      let expectedValue: string | void = originalValue;
      const values = getSourcesFromSrcset(originalValue);
      values.forEach((value) => {
        if (!this.isURLOfCacheableOrigin(value)) return;
        promises.push(
          this.whenReady(value).then((status) => {
            const isLoaded = status.status === 'loaded';
            if (!isLoaded) return; // failed to load asset

            const attributeUnchanged =
              node.getAttribute(attribute) === expectedValue;

            if (!attributeUnchanged) return; // attribute was changed since we started loading the asset

            expectedValue = updateSrcset(node, value, status.url);
          }),
        );
      });
    } else {
      // In live mode we removes the attribute while it loads so it doesn't show the broken image icon
      if (this.liveMode && nodeId > 0) {
        let hijackedAttributes = this.nodeIdAttributeHijackedMap.get(nodeId);
        if (!hijackedAttributes) {
          hijackedAttributes = new Map();
          this.nodeIdAttributeHijackedMap.set(nodeId, hijackedAttributes);
        }
        hijackedAttributes.set(attribute, originalValue);
        if (node.tagName === 'IMG' && attribute === 'src') {
          node.setAttribute('src', '//:0');
        } else {
          node.removeAttribute(attribute);
        }
      }

      promises.push(
        this.whenReady(originalValue).then((status) => {
          const isLoaded = status.status === 'loaded';
          if (!isLoaded) return; // failed to load asset

          const attributeUnchanged = this.liveMode
            ? originalValue ===
              this.nodeIdAttributeHijackedMap.get(nodeId)?.get(attribute)
            : node.getAttribute(attribute) === originalValue;

          if (!attributeUnchanged) return; // attribute was changed since we started loading the asset

          // TODO: use setAttributeNS for svg, see rrdom's diff for example
          node.setAttribute(attribute, status.url);
        }),
      );
    }

    return Promise.all(promises);
  }

  public reset(config?: captureAssetsParam | undefined): void {
    this.config = config;
    this.originalToObjectURLMap.forEach((objectURL) => {
      URL.revokeObjectURL(objectURL);
    });
    this.originalToObjectURLMap.clear();
    this.loadingURLs.clear();
    this.failedURLs.clear();
    this.nodeIdAttributeHijackedMap.clear();
    this.callbackMap.forEach((callbacks) => {
      while (callbacks.length > 0) {
        const cb = callbacks.pop();
        if (cb) cb({ status: 'reset' });
      }
    });
    this.callbackMap.clear();
  }
}
