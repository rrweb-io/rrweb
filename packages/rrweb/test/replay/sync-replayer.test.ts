import { printRRDom } from 'rrdom';
import { SyncReplayer } from '../../src/replay/sync-replayer';
import { EventType, eventWithTime } from '@rrweb/types';
import { events as mutationEvents } from '../events/node-mutation';

describe('A synchronous replayer purely built with RRDom', () => {
  it('should play mutation events synchronously', () => {
    const events = mutationEvents as eventWithTime[];
    const replayer = new SyncReplayer(events);
    replayer.play(({ event, currentTime }) => {
      if (event.type === EventType.FullSnapshot) {
        expect(
          printRRDom(replayer.virtualDom, replayer.getMirror()),
        ).toMatchSnapshot(`Full Snapshot @ ${currentTime}`);
      } else if (event.type === EventType.IncrementalSnapshot) {
        expect(
          printRRDom(replayer.virtualDom, replayer.getMirror()),
        ).toMatchSnapshot(`Incremental Snapshot @ ${currentTime}`);
      }
    });
  });
});
