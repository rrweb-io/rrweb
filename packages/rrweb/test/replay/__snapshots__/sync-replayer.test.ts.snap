// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A synchronous replayer purely built with RRDom should play mutation events synchronously: Full Snapshot @ 100 1`] = `
"1 RRDocument
  -1 RRDocumentType
  2 HTML 
    3 HEAD 
      4 META charset=\\"utf-8\\" 
      5 RRText text=\\"    \\\\n  \\"
    6 BODY 
      16 RRText text=\\"\\\\n    \\"
      17 DIV id=\\"container\\" 
        18 RRText text=\\"\\\\nempty\\\\n\\"
      19 RRText text=\\"\\\\n  \\\\n  \\"
"
`;

exports[`A synchronous replayer purely built with RRDom should play mutation events synchronously: Incremental Snapshot @ 1000 1`] = `
"1 RRDocument
  -1 RRDocumentType
  2 HTML 
    3 HEAD 
      4 META charset=\\"utf-8\\" 
      5 RRText text=\\"    \\\\n  \\"
    6 BODY 
      16 RRText text=\\"\\\\n    \\"
      17 DIV id=\\"container\\" 
        34 DIV 
          35 RRText text=\\"1\\"
        33 DIV 
          36 RRText text=\\"2\\"
        32 DIV 
          37 RRText text=\\"3\\"
        31 DIV 
          38 RRText text=\\"4\\"
        29 DIV 
          30 RRText text=\\"5\\"
      19 RRText text=\\"\\\\n  \\\\n  \\"
"
`;

exports[`A synchronous replayer purely built with RRDom should play mutation events synchronously: Incremental Snapshot @ 2000 1`] = `
"1 RRDocument
  -1 RRDocumentType
  2 HTML 
    3 HEAD 
      4 META charset=\\"utf-8\\" 
      5 RRText text=\\"    \\\\n  \\"
    6 BODY 
      16 RRText text=\\"\\\\n    \\"
      17 DIV id=\\"container\\" 
        29 DIV 
          30 RRText text=\\"5\\"
        31 DIV 
          38 RRText text=\\"4\\"
        32 DIV 
          37 RRText text=\\"3\\"
        33 DIV 
          36 RRText text=\\"2\\"
        34 DIV 
          35 RRText text=\\"1\\"
      19 RRText text=\\"\\\\n  \\\\n  \\"
"
`;

exports[`A synchronous replayer purely built with RRDom should play mutation events synchronously: Incremental Snapshot @ 3000 1`] = `
"1 RRDocument
  -1 RRDocumentType
  2 HTML 
    3 HEAD 
      4 META charset=\\"utf-8\\" 
      5 RRText text=\\"    \\\\n  \\"
    6 BODY 
      16 RRText text=\\"\\\\n    \\"
      17 DIV id=\\"container\\" 
      19 RRText text=\\"\\\\n  \\\\n  \\"
"
`;
